src/core/execute.c

```diff
@@ -3690,8 +3717,17 @@ static int exec_child(
         /* In case anything used libc syslog(), close this here, too */
         closelog();
 
-        n_fds = n_socket_fds + n_storage_fds;
-        r = close_remaining_fds(params, runtime, dcreds, user_lookup_fd, socket_fd, params->exec_fd, fds, n_fds);
+        int keep_fds[n_fds + 1];
+        memcpy_safe(keep_fds, fds, n_fds * sizeof(int));
+        n_keep_fds = n_fds;
+
+        r = add_shifted_fd(keep_fds, ELEMENTSOF(keep_fds), &n_keep_fds, params->exec_fd, &exec_fd);
```

Comment:
wasn't sizeof() of VLA particularly ugly, because needs eval at runtime? wasn't that something to avoid? i don#t remember the details, but i thought there was somehting

Reply:
I tried to check this. VLA allocation is indeed quite expensive, approx. 60 instructions in unoptimized mode. But with `a[n + 1]` there isn't much difference between `ELEMENTSOF(a)` and just `n + 1`. In both in unoptimized mode gcc does some calculation and pushes to the right register. (There is shl and shr with ELEMENTSOF, which is probably more expensive than the add which is done in the other case, but I don't think we need to go into this level of microoptimization.)

With optimization though, gcc replaces the VLA with a simple %rsp adjustment and the ELEMENTSOF with a single move. So I think there is no need to change the code.

Reply:
FTR, test case:
```
#include <stdio.h>
#include <string.h>

#define VOID_0 ((void*)0)

#define ELEMENTSOF(x)                                                   \
        (__builtin_choose_expr(                                         \
                !__builtin_types_compatible_p(typeof(x), typeof(&*(x))), \
                sizeof(x)/sizeof((x)[0]),                               \
                VOID_0))

int foo(unsigned a_n) {
        int b[a_n + 1];

        explicit_bzero(b, a_n + 1);

        return ELEMENTSOF(b);
}
```